lib/cassandra/protocol/cql_protocol_handler.rb

```diff
@@ -275,21 +276,64 @@ def time_out!
           end
         end
 
+        def maybe_start_timer
+          # This is more complicated than one would expect. First, we want to start a timer
+          # if a timeout is set. But there is a race condition where send_request creates
+          # a fresh promise and adds it to @promises, but another thread handles a socket
+          # closure event and fails all known promises. When a promise fails, we want to cancel
+          # the timer, if set. So, we synchronize access to @timer to be sure we don't set up
+          # and cancel the timer at the same time. However, if promise.fail runs first, there
+          # will be no timer to cancel, and then when maybe_start_timer gets called in the other
+          # thread, it'll create a timer on a promise that no one is going to action on going forward.
+          # So, that leads to leaking the timer until it times out. To avoid this, we want to
+          # check that the future of the promise isn't completed before starting the timer.
+          # Finally, we cancel a potentially existing timer before scheduling a new one. That's
+          # just defensive programming -- that scenario shouldn't be possible right now.
+          if @timeout
```

Comment:
I'd rewrite this as the following:
```ruby
return if @timeout.nil?
return if @future.completed?

if @timer.nil?
  @lock.synchronize do
    if @timer.nil?
      return if @future.completed?
      @timer = @scheduler.schedule_timer(@timeout)
      @timer.on_value { time_out! }
    end
  end
end
```

This is the equivalent of java's volatile read/lock/read paradigm since jruby variables are actually volatile fields. We also avoid some begin/ensure ceremony by using `Lock#synchronize`